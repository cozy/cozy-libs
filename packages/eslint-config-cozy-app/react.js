'use strict'

const basics = require('./basics')

module.exports = {
  plugins: basics.plugins.concat(['react-hooks']),
  extends: basics.extends.concat(['plugin:react/recommended']),
  parser: basics.parser,
  parserOptions: { ecmaFeatures: { jsx: true } },
  env: basics.env,
  settings: { react: { version: 'latest' } },
  rules: Object.assign({}, basics.rules, {
    'react/prop-types': 'off',
    'react-hooks/rules-of-hooks': 'error',
    'react-hooks/exhaustive-deps': 'warn'
  })
}
