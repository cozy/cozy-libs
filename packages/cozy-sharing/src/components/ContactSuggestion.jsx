import React from 'react'
import PropTypes from 'prop-types'
import get from 'lodash/get'
import { models } from 'cozy-client'
const ContactModel = models.contact
import { translate } from 'cozy-ui/transpiled/react/I18n'
import Avatar from 'cozy-ui/transpiled/react/Avatar'

import styles from './recipient.styl'

import { Contact, Group, getDisplayName, getInitials } from '../models'
import Identity from './Identity'

export const ContactSuggestion = ({ contactOrGroup, contacts, t }) => {
  let avatarText, name, details
  if (contactOrGroup._type === Group.doctype) {
    name = contactOrGroup.name
    const membersCount = contacts
      .reduce((total, contact) => {
        if (
          get(contact, 'relationships.groups.data', [])
            .map(group => group._id)
            .includes(contactOrGroup._id)
        ) {
          return total + 1
        }

        return total
      }, 0)
      .toString()
    details = t('Share.members.count', {
      smart_count: membersCount
    })
    avatarText = 'G'
  } else {
    name = getDisplayName(contactOrGroup)
    avatarText = getInitials(contactOrGroup)
    details = ContactModel.getPrimaryCozy(contactOrGroup)
  }

  return (
    <div className={styles['recipient']}>
      <Avatar text={avatarText} size="small" />
      <Identity name={name} details={details} />
    </div>
  )
}

ContactSuggestion.propTypes = {
  contactOrGroup: PropTypes.oneOfType([Contact.propType, Group.propType])
    .isRequired,
  contacts: PropTypes.arrayOf(Contact.propType),
  t: PropTypes.func.isRequired
}

export default translate()(ContactSuggestion)
