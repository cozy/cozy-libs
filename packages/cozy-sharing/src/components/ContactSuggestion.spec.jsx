import { render, screen } from '@testing-library/react'
import React from 'react'

import { ContactSuggestion } from './ContactSuggestion'
import AppLike from '../../test/AppLike'

describe('ContactSuggestion component', () => {
  const setup = ({ contactOrGroup }) => {
    return render(<ContactSuggestion contactOrGroup={contactOrGroup} />, {
      wrapper: AppLike
    })
  }

  it('should display contact suggestion for a contact', () => {
    const jonSnow = {
      _id: 'f3a4e501-abbd',
      fullname: 'Jon Snow',
      name: {
        givenName: 'Jon',
        familyName: 'Snow'
      },
      cozy: [
        {
          primary: true,
          url: 'https://jonsnow.mycozy.cloud'
        }
      ],
      email: [
        {
          address: 'jon.snow@email.com',
          type: 'primary'
        }
      ],
      _type: 'io.cozy.contacts'
    }
    setup({ contactOrGroup: jonSnow })
    expect(screen.getByText('Jon Snow')).toBeInTheDocument()
    expect(screen.getByText('https://jonsnow.mycozy.cloud')).toBeInTheDocument()
  })

  it('should display contact suggestion for a contact without fullname (for example created by a share modal)', () => {
    const jonSnow = {
      _id: 'f3a4e501-abbd',
      cozy: [],
      email: [
        {
          address: 'jon.snow@email.com',
          type: 'primary'
        }
      ],
      name: undefined,
      _type: 'io.cozy.contacts'
    }
    setup({ contactOrGroup: jonSnow })
    expect(screen.getByText('jon.snow@email.com')).toBeInTheDocument()
  })

  it('should display contact suggestion for a group', () => {
    const theNightsWatch = {
      _id: '610718e6-2d7a',
      name: "The Night's Watch",
      _type: 'io.cozy.contacts.groups',
      members: [
        {
          _id: 'f3a4e501-abbd',
          fullname: 'Jon Snow'
        },
        {
          _id: 'ab23a451-tzas',
          fullname: 'Samwell Tarly'
        }
      ]
    }
    setup({ contactOrGroup: theNightsWatch })
    expect(screen.getByText("The Night's Watch")).toBeInTheDocument()
    expect(screen.getByText('2 members')).toBeInTheDocument()
  })
})
