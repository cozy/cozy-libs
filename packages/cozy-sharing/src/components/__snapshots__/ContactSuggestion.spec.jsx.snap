// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContactSuggestion component should display contact suggestion for a contact 1`] = `
<ContactSuggestion
  contactOrGroup={
    Object {
      "_id": "f3a4e501-abbd",
      "_type": "io.cozy.contacts",
      "cozy": Array [
        Object {
          "primary": true,
          "url": "https://jonsnow.mycozy.cloud",
        },
      ],
      "email": Array [
        Object {
          "address": "jon.snow@email.com",
          "type": "primary",
        },
      ],
      "fullname": "Jon Snow",
      "name": Object {
        "familyName": "Snow",
        "givenName": "Jon",
      },
    }
  }
  contacts={
    Array [
      Object {
        "_id": "f3a4e501-abbd",
        "_type": "io.cozy.contacts",
        "cozy": Array [
          Object {
            "primary": true,
            "url": "https://jonsnow.mycozy.cloud",
          },
        ],
        "email": Array [
          Object {
            "address": "jon.snow@email.com",
            "type": "primary",
          },
        ],
        "fullname": "Jon Snow",
        "name": Object {
          "familyName": "Snow",
          "givenName": "Jon",
        },
      },
    ]
  }
  t={[MockFunction]}
>
  <div
    className="recipient"
  >
    <Avatar
      className=""
      disabled={false}
      icon="people"
      image=""
      size="small"
      style={Object {}}
      text="JS"
    >
      <div
        className="styles__c-avatar___PpDI- styles__c-avatar--text___2dvna"
        style={
          Object {
            "--circleSize": "32px",
            "backgroundColor": "var(--lavender)",
          }
        }
      >
        <span
          className="styles__c-avatar-initials___310qC"
        >
          JS
        </span>
      </div>
    </Avatar>
    <Identity
      details="https://jonsnow.mycozy.cloud"
      name="Jon Snow"
    >
      <div
        className="recipient-idents u-ml-1"
      >
        <div
          className="recipient-user"
        >
          Jon Snow
        </div>
        <div
          className="recipient-details"
        >
          https://jonsnow.mycozy.cloud
        </div>
      </div>
    </Identity>
  </div>
</ContactSuggestion>
`;

exports[`ContactSuggestion component should display contact suggestion for a contact without fullname (for example created by a share modal) 1`] = `
<ContactSuggestion
  contactOrGroup={
    Object {
      "_id": "f3a4e501-abbd",
      "_type": "io.cozy.contacts",
      "cozy": Array [],
      "email": Array [
        Object {
          "address": "jon.snow@email.com",
          "type": "primary",
        },
      ],
      "name": undefined,
    }
  }
  contacts={
    Array [
      Object {
        "_id": "f3a4e501-abbd",
        "_type": "io.cozy.contacts",
        "cozy": Array [],
        "email": Array [
          Object {
            "address": "jon.snow@email.com",
            "type": "primary",
          },
        ],
        "name": undefined,
      },
    ]
  }
  t={[MockFunction]}
>
  <div
    className="recipient"
  >
    <Avatar
      className=""
      disabled={false}
      icon="people"
      image=""
      size="small"
      style={Object {}}
      text="J"
    >
      <div
        className="styles__c-avatar___PpDI- styles__c-avatar--text___2dvna"
        style={
          Object {
            "--circleSize": "32px",
            "backgroundColor": "var(--puertoRico)",
          }
        }
      >
        <span
          className="styles__c-avatar-initials___310qC"
        >
          J
        </span>
      </div>
    </Avatar>
    <Identity
      details="-"
      name="jon.snow@email.com"
    >
      <div
        className="recipient-idents u-ml-1"
      >
        <div
          className="recipient-user"
        >
          jon.snow@email.com
        </div>
        <div
          className="recipient-details"
        >
          -
        </div>
      </div>
    </Identity>
  </div>
</ContactSuggestion>
`;

exports[`ContactSuggestion component should display contact suggestion for a group 1`] = `
<ContactSuggestion
  contactOrGroup={
    Object {
      "_id": "610718e6-2d7a",
      "_type": "io.cozy.contacts.groups",
      "name": "The Night's Watch",
    }
  }
  contacts={
    Array [
      Object {
        "_id": "f3a4e501-abbd",
        "_type": "io.cozy.contacts",
        "email": Array [
          Object {
            "address": "jon.snow@email.com",
            "type": "primary",
          },
        ],
        "fullname": "Jon Snow",
        "name": Object {
          "familyName": "Snow",
          "givenName": "Jon",
        },
        "relationships": Object {
          "groups": Object {
            "data": Array [
              Object {
                "_id": "610718e6-2d7a",
                "_type": "io.cozy.contacts.groups",
              },
            ],
          },
        },
      },
      Object {
        "_id": "42dc490a-7878",
        "_type": "io.cozy.contacts",
        "email": Array [
          Object {
            "address": "cersei@email.com",
            "type": "primary",
          },
        ],
        "fullname": "Cersei Lannister",
        "name": Object {
          "familyName": "Lannister",
          "givenName": "Cersei",
        },
        "relationships": Object {
          "groups": Object {
            "data": Array [],
          },
        },
      },
      Object {
        "_id": "b5bed853-93da",
        "_type": "io.cozy.contacts",
        "email": Array [
          Object {
            "address": "samwell@email.com",
            "type": "primary",
          },
        ],
        "fullname": "Samwell Tarly",
        "name": Object {
          "familyName": "Tarly",
          "givenName": "Samwell",
        },
        "relationships": Object {
          "groups": Object {
            "data": Array [
              Object {
                "_id": "610718e6-2d7a",
                "_type": "io.cozy.contacts.groups",
              },
            ],
          },
        },
      },
    ]
  }
  t={[MockFunction]}
>
  <div
    className="recipient"
  >
    <Avatar
      className=""
      disabled={false}
      icon="people"
      image=""
      size="small"
      style={Object {}}
      text="G"
    >
      <div
        className="styles__c-avatar___PpDI- styles__c-avatar--text___2dvna"
        style={
          Object {
            "--circleSize": "32px",
            "backgroundColor": "var(--purpley)",
          }
        }
      >
        <span
          className="styles__c-avatar-initials___310qC"
        >
          G
        </span>
      </div>
    </Avatar>
    <Identity
      details="2 members"
      name="The Night's Watch"
    >
      <div
        className="recipient-idents u-ml-1"
      >
        <div
          className="recipient-user"
        >
          The Night's Watch
        </div>
        <div
          className="recipient-details"
        >
          2 members
        </div>
      </div>
    </Identity>
  </div>
</ContactSuggestion>
`;
